### ISSUE TITLE
Conversion between the same type with different const generic values

### ISSUE THREAD

**Issue by Progdrasil** [rust-lang/rust#86978]
I have a struct which uses const generics for defining some of its properties. 
I'd like to use the standard `From` and `Into` traits for converting from one struct to another with different const values.
There exists a non-fallible equation for the conversion.
However the default `impl<T> From<T> for T` is clashing with my implementation.

I tried this code:

```rust
struct Decimal<const X: u32>(/* fields unimportant */);

impl<const X: u32, const Y: u32> From<Decimal<X>> for Decimal<Y> {
    fn from(previous: Decimal<X>) -> Self {
        /*
        mathematical conversions
        */
        Self(/*ommited*/)
    }
}
```

The compiler [outputs](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=049c5e309e049de786b6f9e8ead0f8c4)
```
error[E0119]: conflicting implementations of trait `std::convert::From<Decimal<{_: u32}>>` for type `Decimal<{_: u32}>`
 --> src/lib.rs:3:1
  |
3 | impl<const S1: u32, const S2: u32> From<Decimal<S1>> for Decimal<S2> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `core`:
          - impl<T> From<T> for T;
```

I understand this might fall into specialization territory but I expected this to compile since `T<const X>` and `T<const Y>` should not be the same type. And indeed the seems to agree since the default implementation does not [allow the following](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dcd77ad19570d695ac49865e995c0dac):

```rust
struct Decimal<const S: u32>(/* fields unimportant */);

fn main() {
    let x = Decimal::<7>();
    let y: Decimal<8> = x.into();
}
```

This fails to compile with:
```
error[E0277]: the trait bound `Decimal<8_u32>: From<Decimal<7_u32>>` is not satisfied
 --> src/main.rs:5:27
  |
5 |     let y: Decimal<8> = x.into();
  |                           ^^^^ the trait `From<Decimal<7_u32>>` is not implemented for `Decimal<8_u32>`
  |
  = note: required because of the requirements on the impl of `Into<Decimal<8_u32>>` for `Decimal<7_u32>`

error: aborting due to previous error
```

[quinedot](https://users.rust-lang.org/t/conversion-between-the-same-type-but-different-const-generics/61837/2?u=progdrasil) tried some nightly for the case where `X==Y` but [it also does not work](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=049c5e309e049de786b6f9e8ead0f8c4).

### Meta

[quinedot](https://users.rust-lang.org/t/conversion-between-the-same-type-but-different-const-generics/61837/2?u=progdrasil) tried some nightly for the case where `X==Y` but [it also does not work](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=049c5e309e049de786b6f9e8ead0f8c4).

`rustc --version --verbose`:
```
1.53.0
```

**Comment by BoxyUwU** [comment]
```rust
struct Decimal<const X: u32>(/* fields unimportant */);

impl<const X: u32, const Y: u32> From<Decimal<X>> for Decimal<Y> {
    fn from(previous: Decimal<X>) -> Self {
        /*
        mathematical conversions
        */
        Self(/*ommited*/)
    }
}
```
this code not compiling is expected. It does conflict with the impl for `From<T> for T` as this impl implies `From<Decimal<1>> for Decimal<1>`. `Decimal<1>` and `Decimal<1>` are the same type

**Comment by Progdrasil** [comment]
> this code not compiling is expected. It does conflict with the impl for `From<T> for T` as this impl implies `From<Decimal<1>> for Decimal<1>`. `Decimal<1>` and `Decimal<1>` are the same type

Yes however, even in nightly with the `NotEqual` implementation it still clashes even though it rules out the case where `X==Y`.

**Comment by MithicSpirit** [comment]
I am also running into this problem. It seems to be a special case of
```rust
struct Foo<T>();
impl<T, U> From<Foo<U>> for Foo<T>  // error: conflicting implementations
{ fn from(value: Foo<U>) -> Self { Self() } }
```
where the parameter `<T>` is a const generic rather than a regular type generic. I feel like this more general case should be resolved first, and then that solution can be generalized to const generics (couldn't find an issue for it, but it probably exists, or there is some easy way around it that I'm not aware of).

Maybe it should just be possible to opt out of blanket implementations for specific types? A more complicated, though perhaps more interesting solution would be to do what [Lean](https://lean-lang.org/) does and have priorities on implementations; then, conflicting implementations would be fine as long as they have different priorities.
